### DISCUSSION TITLE
Performance of uWebSockets.js in Bun

### DISCUSSION THREAD

**Discussion by git-matrix** [uNetworking/uWebSockets.js#931] [Q&A] [answered]
Which should have better performance, using ```uWebSockets.js``` in node.js or bun.sh?

**Comment by uNetworkingAB** [comment]
The problem with JavaScript is that it is a script. So any abstraction adds runtime cost. Right now Bun exposes (only) a Fetch-style JS API with Request and Response objects, which obviously is more abstract (GC) than the function-only interface of uWS.js (non-GC).

**Comment by e3dio** [comment]
They should expose the same API as uWS for Http/ws. They don't even have basic settings working, no progress on issue I opened  https://github.com/oven-sh/bun/issues/2988 idleTimeout doesn't work

I will try import uWS into Zig and code server in Zig next, nice low level language

**Reply by mtsewrs** [reply]
You could probably use the capi and ffi in bun directly idk

**Reply by e3dio** [reply]
No need for Bun if you don't use JavaScript just code in Zig, this is not for everyone you have to learn new language

**Reply by uNetworkingAB** [reply]
> They should expose the same API as uWS for Http/ws

That's not for you or me to decide. Bun is not controlled by us and they can do whatever they want. It just happens to be the Fetch style API for now, maybe it will get a lower level interface in the future but for now there is no such need since Bun runs faster than Deno, Node.js out of the box.

As soon as you add just the slightest bit of JS to your app, the difference will plane out since most overhead is lost in JS being JS

**Comment by git-matrix** [comment]
Both ```bun.sh``` & ```node.js```, supports ```uWebSockets.js```. As of now,  ```bun.sh``` provides much better performance on top of ```node.js```.

Is it also true for ```uWebSockets.js```? Will it give me better performance over ```req/sec```?

**Reply by uNetworkingAB** [reply]
I can't interpret your message in any meaningful way but I already answered your first question

**Reply by git-matrix** [reply]
I am not asking for performance between the languages between ```c++``` & ```javascript```, which is ```uWebSockets``` & ```uWebSockets.js```. I am asking for performance of ```uWebSockets.js``` on ```node.js``` & ```bun.sh```.

**Reply by Mupli** [reply]
+1 to the question .

**Reply by uNetworkingAB** [reply]
You are asking but you aren't listening. I already perfectly described the situation, but you don't seem to accept it.

**Reply by git-matrix** [reply, marked as answer]
Now, I once tried to benchmark the performance of ```node.js``` & ```bun.sh```.

Tested using ```autocannon```.
Tested on my 8-year old laptop having 2-core 2GB RAM with ```Windows10``` installed and running 
-  Microsoft Edge with current discussions page open
- Visual Studio Code Insiders
- Microsoft Terminal
- WSL with Ubuntu-latest installed

Results - 
|         | req/sec | Cpu Usage | Memory Usage | 
|---------|---------|-----------|--------------|
| node.js | 18k     | 100%      | 68%          | 
| bun.sh  | 16k     | 100%      | 73%          |

These are tested at 23% Cpu Usage & 65% Memory Usage on idle.

Code -
```javascript
if (process.isBun) {
    Bun.serve({
        fetch: async () => {
            return new Response('OK');
        },
        port: 8081
    });
} else {
    const { App } = require("uWebSockets.js");

    App()
        .get('/', async res => res.cork(() => {
            res.end('OK');
        })).listen(8080, (socket) => {
            if (socket)
                console.log('Server Listening');
        })
};
```

**Reply by uNetworkingAB** [reply]
Right. Again the problem is that JavaScript is a script. So the creation of new Response('OK') has a **dynamic runtime cost**. So even though both Bun and Node.js use the same underlying server, they differ in perf from this abstraction difference. Creating an object in languages with zero-cost abstractions do not add any performance penalty, but here, in script, we add both Garbage Collection overhead and indirection overhead.

**Reply by git-matrix** [reply]
Whenever we develop or build an application, we just don't go only for performance. Flexibility, Simplicity and Readability are also required. If performance is the only key factor, then, Every website hosted could have built their own framework.?

**Comment by Mupli** [comment]
Sorry for additional question in this topic, But does switching to bun will give us better performance? Node has serious amount of libraries, that don't work with bun. Is it really worth the change?

**Comment by uNetworkingAB** [comment]
It's your task to solve your problems, in whatever way, using whatever tools you decide to use. Nobody is going to decide for you, that's not part of this project.
